File: src/main/java/kr/co/bookand/backend/notice/domain/dto/NoticeDto.java
package kr.co.bookand.backend.notice.domain.dto;

import kr.co.bookand.backend.common.domain.DeviceOSFilter;
import kr.co.bookand.backend.common.domain.MemberIdFilter;
import kr.co.bookand.backend.common.domain.Status;
import kr.co.bookand.backend.notice.domain.Notice;
import kr.co.bookand.backend.notice.domain.NoticeType;
import lombok.Builder;

public class NoticeDto {

    public record NoticeResponse(
            Long id,
            String title,
            String content,
            String createdAt,
            String image,
            String targetType,
            String targetNum
    ) {
        public static NoticeResponse of(Notice notice) {
            return new NoticeResponse(
                    notice.getId(),
                    notice.getTitle(),
                    notice.getContent(),
                    notice.getCreatedAt(),
                    notice.getImage(),
                    notice.getDeviceOSFilter().toString(),
                    notice.getMemberIdFilter().toString()
            );
        }
    }

    public record NoticeWebResponse(
            Long id,
            String title,
            String content,
            String image,
            String createdAt,
            String modifiedAt,
            String status,
            String noticeType,
            String targetType,
            String targetNum
    ) {
        public static NoticeWebResponse of(Notice notice) {
            return new NoticeWebResponse(
                    notice.getId(),
                    notice.getTitle(),
                    notice.getContent(),
                    notice.getImage(),
                    notice.getCreatedAt(),
                    notice.getModifiedAt(),
                    notice.getStatus().toString(),
                    notice.getNoticeType().toString(),
                    notice.getDeviceOSFilter().toString(),
                    notice.getMemberIdFilter().toString()
            );
        }
    }

    public record CreateNoticeRequest(
            String title,
            String content,
            String image,
            String noticeType,
            String targetType,
            String targetNum
    ) {
        @Builder
        public CreateNoticeRequest {
        }

        public Notice toEntity() {
            return Notice.builder()
                    .title(title)
                    .content(content)
                    .image(image)
                    .noticeType(NoticeType.valueOf(noticeType))
                    .deviceOSFilter(DeviceOSFilter.valueOf(targetType))
                    .memberIdFilter(MemberIdFilter.valueOf(targetNum))
                    .status(Status.INVISIBLE)
                    .build();
        }
    }

    public record UpdateNoticeRequest(
            String title,
            String content,
            String image,
            String noticeType,
            String targetType,
            String targetNum
    ) {
        @Builder
        public UpdateNoticeRequest {
        }
    }

    public record NoticeMessage(String message) {
    }

    public record NoticeIdResponse(Long id) {
    }
}

File: src/main/kotlin/kr/co/bookand/backend/notice/dto/NoticeDto.kt
package kr.co.bookand.backend.notice.dto

import kr.co.bookand.backend.common.DeviceOSFilter
import kr.co.bookand.backend.common.MemberIdFilter
import kr.co.bookand.backend.common.PageResponse
import kr.co.bookand.backend.common.Status
import kr.co.bookand.backend.notice.model.Notice


data class CreateNoticeRequest(
    val title: String,
    val content: String,
    val image: String,
    val noticeType: String,
    val noticeFilter: NoticeFilter
)

data class NoticeIdResponse(
    val id: Long
)

data class UpdateNoticeRequest(
    val title: String,
    val content: String,
    val image: String,
    val noticeType: String,
    val noticeFilter: NoticeFilter
)

data class NoticeFilter(
    val deviceOS: DeviceOSFilter,
    val memberId: MemberIdFilter
)

data class NoticePageResponse(
    val data: PageResponse<NoticeResponse>
)

data class NoticeResponse(
    val id: Long,
    val title: String,
    val noticeType: String,
    val createdAt: String,
    val displayAt: String?,
    val modifiedAt: String,
    val isDisplayed: Boolean,
    val content: String,
    val image: String,
    val noticeFilter: NoticeFilter
) {
    constructor(notice: Notice) : this(
        id = notice.id,
        title = notice.title,
        noticeType = notice.noticeType.name,
        createdAt = notice.createdAt.toString(),
        displayAt = notice.displayAt?.toString(),
        modifiedAt = notice.modifiedAt.toString(),
        isDisplayed = notice.status == Status.VISIBLE,
        content = notice.content,
        image = notice.image,
        noticeFilter = NoticeFilter(
            deviceOS = notice.deviceOSFilter,
            memberId = notice.memberIdFilter
        )
    )
}